### DISCUSSION TITLE
uv_write_cb not being called after write on async_send

### DISCUSSION THREAD

**Discussion by s2548472** [libuv/libuv#4457] [Q&A] [answered]
I have a multithreaded application where I have several worker threads and a different thread running the uv_loop. I am writing back to the loop using an async_send cb. However, when I do it that way the uv_write_cb doesn't get called immediately, but it seems to be called the next time some other event callback happens on the loop. On the example below, 'on_write_end' is not called but until there is some other event on the loop. 
Is that expected or maybe I am missing something?
```c
void async_write(uv_async_t* async) {
    response_data_t *response_data = async->data;
    
    uv_stream_t *client = response_data->client;
    client_request_data *client_data = client->data;    
    int* buffer = response_data->buffer;
    size_t size = response_data->size;

    uv_buf_t wrbuf = uv_buf_init((char*)buffer,size);
    uv_write_t* write_req = response_data->write_req;
    write_req->data = (void*)client_data->client;  
    
    uv_write(write_req,client,&wrbuf,1,on_write_end);

    client_data->ready_to_write = 1;
}
```

**Comment by bnoordhuis** [comment]
No, that's not expected. 9 out of 10 times it's some kind of thread safety issue in the program but your snippet doesn't provide enough context for me to guess what.

NB: I'm assuming that "immediately" doesn't mean you expect on_write_end to have been called by the time uv_write() returns.

**Reply by s2548472** [reply]
No, I am not assuming that. The issue is that I see the incoming data on the client side being received, and the 'on_write_end' callback gets executed a long time after that. Sometimes, even after the other side closes the connection.  

Could it be related to the fact that the thread calling async_send() was not created with the uv_thread utilities but just with pthread?

**Reply by s2548472** [reply]
I added some logging to my code. This is what I am seeing:
```c
void on_write_finish(uv_write_t *write, int status) {
    /*  Callback for when the TCP write is complete
     *  Not currently using it. 
     * */
    log_str(&tcp_logger, WARNING, TCP_TAG, "%s", "Successfully sent data to client");
}

void async_write(uv_async_t* async) {
    /* Function to perform async write on main loop
     * This function is waken up by async_send thread
     * */
    response_data_t *response_data = async->data;
    const char* WT_TAG = response_data->WT_TAG;
    log_str(&tcp_logger, WARNING, WT_TAG, "%s", "Async_write task started, awaking uv_loop");
    
    uv_stream_t *client = response_data->client;
    client_request_data *client_data = client->data;    
    int* buffer = response_data->buffer;
    size_t size = response_data->size;

    uv_buf_t wrbuf = uv_buf_init((char*)buffer,size);
    uv_write_t* write_req = response_data->write_req;
    write_req->data = (void*)client_data->client;  
    
    log_str(&tcp_logger, WARNING, WT_TAG, "%s %d %s", "Sending ",size," Bytes to client");
    uv_write(write_req,client,&wrbuf,1,on_write_finish);

    client_data->ready_to_write = 1;
}
```
There is no other lock in the uv_loop or anything else that can be locked. I would expect to see the logging from the writing call immediately as soon as the client receives the data. However, I am seeing it after the next read_event.  It seems like the write_cb is only queued after there is a read_cb. Would that make sense? Or should I look for other broken stuff in my code? See an example of the logging I'm seeing:

- W - [Thu Jul 18 21:11:22 2024] - worker_thread_0: Async_write task started, awaking uv_loop
-  W - [Thu Jul 18 21:11:22 2024] - worker_thread_0: Sending  52  Bytes to client
-  I - [Thu Jul 18 21:11:28 2024] - tcp_server: Initialize buffer of size:  65536  Bytes
-  I - [Thu Jul 18 21:11:28 2024] - tcp_server: Read event triggered
- W - [Thu Jul 18 21:11:28 2024] - tcp_server: Client closed the connection
- I - [Thu Jul 18 21:11:28 2024] - tcp_server: Freeing read_cb buffer because of client close
- W - [Thu Jul 18 21:11:28 2024] - tcp_server: Successfully sent data to client

PD. Sorry for the code preview, I couldn't make it more readable

**Reply by bnoordhuis** [reply]
> Could it be related to the fact that the thread calling async_send() was not created with the uv_thread utilities but just with pthread?

No, that should make no difference.

> I would expect to see the logging from the writing call immediately as soon as the client receives the data.

At the risk of being pedantic: as soon as libuv is done _sending_ it - which is not the same as the client _receiving_ it.

> It seems like the write_cb is only queued after there is a read_cb. Would that make sense?

No, they're independent.

If you can put together a minimal but complete example, I can probably tell you what the issue is, but right now there's no guessing.

**Reply by s2548472** [reply]
Thanks for your help. I found the issue: I was reusing the uv_write_t struct for several uv_write operations, which caused a weird behavior with the callback.

**Reply by s2548472** [reply]
At the risk of sounding like an idiot, I need to confess that I fell into this rabbit whole again. Let me try to give the most comprehensive example possible. 

I have a worker thread, and my uv_loop is running on a different thread. The Worker thread processes some data, and when it is finished it sends data back to the uv_loop thread through and async call. 

Here is a sample of what the worker thread does:

```
    for(;;){
         /* Lock the job_queue so no one else writes to it while executing it */
        pthread_mutex_lock(&pending_jobs->queue_mutex);
        while(job == NULL){
            pthread_cond_wait(&pending_jobs->queue_cond,&pending_jobs->queue_mutex);
            job = dequeue_job(pending_jobs);
        }
        pthread_mutex_unlock(&pending_jobs->queue_mutex);
      
        log_str(&wt_loggers[id], ERROR, WT_TAG, "%s%d", "Job received on WT",id); 
        execute_job_q(job);
        log_str(&wt_loggers[id], ERROR, WT_TAG, "%s%d", "Job executed on WT",id);

        /* Send a message out*/
        pthread_mutex_lock(&completed_jobs->queue_mutex);
        enqueue_job(job,completed_jobs);
        log_str(&wt_loggers[id], ERROR, WT_TAG, "%s %d", "Queueing for write on position",completed_jobs->size);
        pthread_mutex_unlock(&completed_jobs->queue_mutex);
        async->data = job->client;
        async_write(async);
        job = NULL;
    }
}

```

The async_write call marshalls the information out, and does a uv_write_ which in turn assings the callback.

```
void on_write_finish(uv_write_t *write, int status) {
    /*  Callback for when the TCP write is complete */
    log_str(&tcp_logger, WARNING, TCP_TAG, "%s", "Successfully sent data to client");
}

void async_write(uv_async_t* async) {
    /* Function to perform async write on main loop*/
    log_str(&tcp_logger, WARNING, TCP_TAG, "%s", "Async_write task started, awaking uv_loop");
    job_queue_t *completed_jobs = loop_info->completed_jobs;
    
    /* Load the client data */
    uv_stream_t *client = async->data;
    client_data_t *client_data = client->data;    
    client_data->response = (int*)malloc(MAX_JOB_SIZE);
    size_t size = 0;

    /* Marshall one job for now */
    pthread_mutex_lock(&completed_jobs->queue_mutex);
    job_q_t *job = dequeue_job(completed_jobs);
    size = marshall_result_q(job,client_data->response);
    pthread_mutex_unlock(&completed_jobs->queue_mutex);

    /* Init structures for syncronus send */
    uv_write_t *req = (uv_write_t*)malloc(sizeof(uv_write_t));
    uv_buf_t wrbuf = uv_buf_init((char*)client_data->response,size);

    log_str(&tcp_logger, WARNING, TCP_TAG, "%s %d %s", "Sending ",(int)size," Bytes to client");
    
    /* This is what actually sends */
    uv_write(req,client,&wrbuf,1,on_write_finish);
}
```

My main problem is that the 'on_write_finish' callback is not reached but until something else happens on the uv_loop (eg. a client disconnects or there is another read_cb event).  I would like to use the callback to free my buffers and other housekeeping operations.  There are no other mutex_locks in the uv_loop thread other than what displayed in the example above which should happen before the uv_write is done and after any subsequent call to it again.

**Reply by saghul** [reply]
Have you considered using uv_queue_work instead? It's integrated into libuv and it allows you to do work in an external thread easily.

**Reply by s2548472** [reply, marked as answer]
I must admit that I am a bit ashamed of the source of the issue, but all seemed to came down for me not calling async events using uv_async_send() mechanism. 

@saghul, yes I've used them and like them very much. I am using this for a producer-consumer model for my MSc thesis and therefore wanted more control on instantiation of threads.
